### ISSUE TITLE
*OMIT not supported 

### ISSUE THREAD

**Issue by GajenderI** [IBM/nodejs-itoolkit#343]
What if the RPG program has a *omit param and I want to specify that in the addParm and also not pass any value to it. 

**Describe the solution you'd like**
omit should be supported 

**Describe alternatives you've considered**
None 

**Additional context**
In the XMLSERVICE documentation we can see that a param option io can have values 'in|out|both|**omit'** but the itoolkit document says io (string()) –  Valid values are in, out, or both.
https://nodejs-itoolkit.readthedocs.io/en/latest/ProgramCall.html#parameterConfig
https://xmlservice.readthedocs.io/en/latest/functions.html#pgm

**Comment by alanseiden** [comment]
We will look into *OMIT since XMLSERVICE supports it.

**Comment by abmusse** [comment]
Yes the documentation needs to be updated to include `*OMIT` as a valid option.
As the code is written passing `*omit` as `io` should still get passed through though.

https://github.com/IBM/nodejs-itoolkit/blob/c4f5e2952cd4eced07f940e704e609008eed15a7/lib/ProgramCall.js#L141

**Comment by GajenderI** [comment]
@alanseiden Thank you for considering this.
@abmusse Thank you for the pointing that out, I will try this and get back to you.

**Comment by alanseiden** [comment]
@GajenderI Please try `omit` again, if you still have your program available, and let us know how it goes. If you get an error, please tell us the specific error. We may still change the documentation to allow the `omit` option.

**Comment by alanseiden** [comment]
We have created a PR to update the documentation after testing 'omit' in XMLSERVICE: https://github.com/IBM/nodejs-itoolkit/pull/346

**Comment by alanseiden** [comment]
@GajenderI here is what we found so far:
- the ZZOMIT unit test works correctly.
- Your program test does not fully work (we replicated your results). The parameter comes in blank (good), but is not equal to a null pointer (bad).

The difference may be that ZZOMIT is a service program. We are looking at why this may matter. XMLSERVICE has a different code path for service program procedures.

@GajenderI If you have time, try making your program into a service program with a procedure having the *OMIT and let us know if that makes a difference.

**Comment by alanseiden** [comment]
We should create a new issue about your specific test rather than adding to this closed issue that was about documentation.

**Comment by sagartyagi121** [comment]
Thank you very much @alanseiden for taking the effort. I will test it by making it a service program and let me know how else I can help in this.

**Comment by sagartyagi121** [comment]
@amagid please have a look at this issue.

**Comment by JonFP** [comment]
> The difference may be that ZZOMIT is a service program. We are looking at why this may matter. XMLSERVICE has a different code path for service program procedures.
>

The mechanism the system uses for program parameters and procedure parameters is _radically_ different Alan so this is hardly surprising.  Parms for procedures (regardless of whether local/bound/or service program) are pushed on the stack. So to get the effect of *Omit a null pointer has to be pushed on the stack.  For program calls the parm pointers are passed to the system and retrieved basically by calling an internal API. So radically different code paths.

**Comment by kadler** [comment]
> We should create a new issue about your specific test rather than adding to this closed issue that was about documentation.

If this can be replicated outside of nodejs-itoolkit using raw XML to XMLSERVICE, then it's likely an XMLSERVICE issue and an issue should be opened there.

**Comment by brandonp42** [comment]
@kadler It is a XMLSERVICE issue and I'm hoping to submit a PR for it later this week, but it might only be for the mode="opm" path. The mode="ile" code path that does PASE stuff is quite a bit more complicated at first glance.

**Comment by brandonp42** [comment]
I just realized that I'm not seeing a way with the nodejs-itoolkit to specify that a program call should be mode="opm" vs mode="ile". So if I only fix the mode="opm" path in XMLSERVICE then there's still going to be a problem because it seems to default to the mode="ile" handling. If I'm correct would @sagartyagi121 or @amagid be willing to submit a PR adding that ability here?

**Comment by JonFP** [comment]
As far as program calls go Brandon there should be no difference between the two. I'm actually surprised that there would be a different code path. A program call is a program call is a program call. ILE or OPM I don't think it matters.  ILE diferentaion should only be needed for procedure calls.

Can you point me to the source you are studying and I'll see if I can make anything of it.  I last studied XMLSERVICE a LONG time ago and it has been re-hosted since then,


Jon Paris

***@***.***

**Comment by brandonp42** [comment]
@JonFP I'm not in the debugger right now but here's my recollection from Friday. It is pretty challenging to follow the code path through and ultimately I had to just walk through the debugger to see it.

`<pgm mode="ile"> *PGM` - https://github.com/IBM/xmlservice/blob/master/src/plugpase.rpgle#L1526
`<pgm mode="opm"> *PGM` - https://github.com/IBM/xmlservice/blob/master/src/plugile.rpgle#L4293
`<pgm mode="ile"> *SRVPGM` - https://github.com/IBM/xmlservice/blob/master/src/plugpase.rpgle#L1634
`<pgm mode="opm> *SRVPGM` - https://github.com/IBM/xmlservice/blob/master/src/plugile.rpgle#L4366

You can see that for the mode="ile" path via PASE it's using Qp2CallPase for both *PGM and *SRVPGM objects, but mode="opm" is using _CALLPGMV for *PGM objects and a semi-generic prototype with procedure pointer for *SRVPGM function calls. I suspect the prototype might not work correctly for parameters passed by value but I haven't tried to test that yet.

Also from the XMLSERVICE documentation at https://xmlservice.readthedocs.io/en/latest/functions.html#xmlservice-call-pgm note that the ILE vs OPM doesn't seem to be what you might think at face value (it wasn't what I thought):

```
*  mode
*     ile     - ILE and PASE memory (default)
*     opm     - ILE only memory (PASE can not view)
```

**Comment by JonFP** [comment]
I confess that I'm having a huge problem trying to understand the documentation (or rather the somewhat odd English in which much of it is written).

I really don't understand how this "but mode="opm" is using _CALLPGMV for *PGM objects and a semi-generic prototype with procedure pointer for *SRVPGM function calls." makes any sense. Not saying you are wrong that is what it says - but you can't call a service program from OPM so what the hell do ILE and OPM mean in this context? As you say it clearly doesn't have the obvious meaning and the documentation seems to be no help at all.  The source itself is not commented pin any meaningful way as far as I can tell.


Jon Paris

***@***.***

**Comment by brandonp42** [comment]
Honestly I don't understand why there's the two code paths either. There seems to be reasons based on internal operating system architecture that I'm not sure if anybody outside of IBM really knows well.

At the end of the day though, fixing the PLUGILE ilePGM() & ileSRV() for *OMIT parameters seems somewhat straightforward since it just needs to go through argv backwards instead of forwards. I'm trying to confirm if you can count on all of those pointers having been reliably initialized since it isn't being done in the PLUGILE module. If the memory is only allocated one by one then going backwards through argv in a simplistic fashion won't work. Or maybe there's a count of parameters somewhere that can be grabbed instead of walking through argv.

Fixing *OMIT for PLUGPASE *PGM objects might be possible for me, not sure if I'm going to try yet. It does seem to work for *SRVPGM objects already. I'm almost certainly not going to try to fix the  %PARMS issue in the PLUGPASE code though. I strongly think IBM needs to address that part.

**Comment by alanseiden** [comment]
Tony meant OPM mode to mean using the pre-6.1 memory model. We called it the "v5r4" flag in the PHP toolkit :)

**Comment by alanseiden** [comment]
I found an old email discussion with Tony about *OMIT and OPM / "v5r4" mode. I'll forward it to Kevin, Brandon, Jon, and anyone else interested. It refers to a discussion on the old Zend forum that Zend later deleted, but you can get the main idea from the emails.

**Comment by brandonp42** [comment]
Oh cool, I will look for that.

FWIW there's also the newer JSON based iService (https://github.com/IBM/iService-ibmi) but I don't think it supports all the features yet.

**Comment by alanseiden** [comment]
And "Technically, mode='opm' uses non-teraspace memory compatible both ILE and OPM calls (theory)." I'll forward that one as well.

**Comment by brandonp42** [comment]
I've submitted my PR for mode="opm" at https://github.com/IBM/xmlservice/pull/64

**Comment by kadler** [comment]
Yeah, XMLSERVICE code and options are very obtuse and confusing with overloaded meanings. I did not realize there was this split in opm/ile behavior. We consciously _do not_ make visible all the XMLSERVICE options because of this problem. I do recall having discussions with @abmusse about mode=opm/ile and deciding not to allow it. We might need to do something about it, though I'd prefer to make it somewhat sensibly instead of just passing through the XMLSERVICE option to the user.

**Comment by alanseiden** [comment]
I'll search my notes for one or two other situations where OPM mode is useful.

**Comment by kadler** [comment]
I have no idea why Tony made the decision to call through PASE for program calls when CALLPGMV works great from ILE. In fact, [_PGMCALL](https://www.ibm.com/docs/en/i/7.2?topic=ssw_ibm_i_72/apis/pase__pgmcall.html) is **deficient** in that it requires a null-terminated list of arguments, meaning you _cannot pass null pointer arguments in the middle of the list_ (`*OMIT` parameters won't work!).

Calling in to PASE to call ILE service programs _does_ make sense, because there is no generic way to dynamically call all types of ILE service program functions _from within ILE_. There is [QZRUCLSP](https://www.ibm.com/docs/api/v1/content/ssw_ibm_i_74/apis/qzruclsp.htm) but it only supports up to 248 arguments, which can be either a pointer or a 32-bit integer by value. There's no support for passing floats, decfloats, or any other size integers. However, basically any function can be called using `_ILECALLX` in PASE.

**Comment by brandonp42** [comment]
@kadler is there any way to get _ILECALLX to generate the operational descriptor that RPG needs for %PARMS? See the issue I opened for more details at https://github.com/IBM/xmlservice/issues/62

I'd be happy to add more commentary there if it would help, I've learned a lot more about the XMLSERVICE code this week since creating that.

**Comment by kadler** [comment]
> @kadler is there any way to get _ILECALLX to generate the operational descriptor that RPG needs for %PARMS?

No, this is not supported in PASE. There's an Idea to add support, though: https://ibm-power-systems.ideas.ibm.com/ideas/IBMI-I-3080
